akka-javasdk/src/main/scala/akka/javasdk/impl/workflow/ReflectiveWorkflowRouter.scala

```diff
@@ -4,63 +4,212 @@
+  }
+
+  final case class WorkflowStepNotSupported(stepName: String) extends RuntimeException {
+    override def getMessage: String = stepName
+  }
+}
 
 /**
  * INTERNAL API
  */
 @InternalApi
 class ReflectiveWorkflowRouter[S, W <: Workflow[S]](
-    override protected val workflow: W,
-    commandHandlers: Map[String, CommandHandler])
-    extends WorkflowRouter[S, W](workflow) {
+    val workflow: W,
+    commandHandlers: Map[String, CommandHandler],
+    serializer: JsonSerializer) {
+
+  private var state: Option[S] = None
+  private var workflowFinished: Boolean = false
+
+  private def stateOrEmpty(): S = state match {
+    case None =>
+      val emptyState = workflow.emptyState()
+      // null is allowed as emptyState
+      state = Some(emptyState)
+      emptyState
+    case Some(state) =>
+      state
+  }
+
+  /** INTERNAL API */
+  // "public" api against the impl/testkit
+  def _internalSetInitState(s: Any, finished: Boolean): Unit = {
+    if (!workflowFinished) {
+      state = Some(s.asInstanceOf[S])
+      workflowFinished = finished
+    }
+  }
 
   private def commandHandlerLookup(commandName: String) =
     commandHandlers.getOrElse(
       commandName,
       throw new HandlerNotFoundException("command", commandName, commandHandlers.keySet))
 
-  override def handleCommand(
+  /** INTERNAL API */
+  // "public" api against the impl/testkit
+  final def _internalHandleCommand(
       commandName: String,
-      state: S,
-      command: Any,
-      commandContext: CommandContext): Workflow.Effect[_] = {
-
-    workflow._internalSetCurrentState(state)
-    val commandHandler = commandHandlerLookup(commandName)
-
-    val scalaPbAnyCommand = command.asInstanceOf[ScalaPbAny]
-    if (AnySupport.isJson(scalaPbAnyCommand)) {
-      // special cased component client calls, lets json commands trough all the way
-      val methodInvoker = commandHandler.getSingleNameInvoker()
-      val deserializedCommand =
-        CommandSerialization.deserializeComponentClientCommand(methodInvoker.method, scalaPbAnyCommand)
-      val result = deserializedCommand match {
-        case None          => methodInvoker.invoke(workflow)
-        case Some(command) => methodInvoker.invokeDirectly(workflow, command)
+      command: BytesPayload,
+      context: CommandContext,
+      timerScheduler: TimerScheduler): CommandResult = {
+
+    val commandEffect =
+      try {
+        workflow._internalSetTimerScheduler(Optional.of(timerScheduler))
+        workflow._internalSetCommandContext(Optional.of(context))
+        workflow._internalSetCurrentState(stateOrEmpty())
+
+        val commandHandler = commandHandlerLookup(commandName)
+
+        // Commands can be in three shapes:
+        // - BytesPayload.empty - there is no real command, and we are calling a method with arity 0
+        // - BytesPayload with json - we deserialize it and call the method
+        // - BytesPayload with Proto encoding - we deserialize using InvocationContext
```

Comment:
I commented on WorkflowRouter, but is now gone. 

So, repeating here...

@patriknw, this is related with the `NoCommandPayload`. 
It was a bit obscure to use a no command with json content type and have it being used here. 

Instead, I'm using `BytesPayload.empty` instead. See PR in runtime. 

The workflow can also have an empty state set to `null` that needs to be mapped to a `BytesPayload`. At least, in previous impl, it was being mapped to `PbAny.defaultInstance`.